### PR TITLE
ISSUES-3105 fix some bugs in the predicate optimizer

### PR DESCRIPTION
**Pull request by zhang2014**
- [x] Don't optimize arrayJoin (#3105)
- [x] Support optimize ``SELECT * FROM (SELECT 1) WHERE `1` = 1`` (#3105)
- [x] Fix PredicateOptimizer's error when optimization view with select partial columns (#3111)
- [x] Normalize subqueries before predicate expression optimization (https://travis-ci.org/yandex/ClickHouse/builds/427635053#L8276)

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

File: dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp [hunk 1]
```diff
@@ -210,7 +228,12 @@ void PredicateExpressionsOptimizer::cloneOuterPredicateForInnerPredicate(
         for (auto projection : projection_columns)
         {
             if (require.second == projection.second)
-                require.first->name = projection.first->getAliasOrColumnName();
+            {
+                ASTPtr & ast = projection.first;
+                if (ast->tryGetAlias().empty())
+                    ast->setAlias(ast->getAliasOrColumnName());
```
Comment (zhang2014) [inline dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp]: https://github.com/yandex/ClickHouse/issues/3105#issuecomment-420477177
I tried to solve this problem by set alias.

Comment (alexey-milovidov) [reply dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp]: Ok.

**Comment by zhang2014** [comment]
@alexey-milovidov  Done.

**Comment by alexey-milovidov** [comment]
:+1:

**Pull request merged**